README.md
<a name="top"></a>

<h1 align="center">
  Termcs
</h1>

<h4 align="center">Terminal crypto screener written in Python</h4>

<p align="center">
  <a href="https://pypi.org/project/termcs/">
    <img alt="PyPI" src="https://img.shields.io/pypi/v/termcs">
  </a>
  <a href="https://www.python.org/downloads/">
    <img alt="PyPI - Python Version" src="https://img.shields.io/pypi/pyversions/termcs">
  </a>
  <a href="https://black.readthedocs.io/en/stable/">
    <img alt="Black" src="https://img.shields.io/badge/code_style-black-black">
  </a>
  <a href="https://opensource.org/licenses/MIT">
    <img alt="MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg?style=flat">
  </a>
</p>

<p align="center">
  <a href="#features">Features</a> •
  <a href="#install">Install</a> •
  <a href="#usage">Usage</a>
</p>

<p align="center">
  <img align="center" src="https://raw.githubusercontent.com/Riyum/termcs/master/imgs/demo.png" />
</p>

<details>
  <summary>Table of Contents</summary>
  <ol>
    <li><a href="#features">Features</a></li>
    <li><a href="#install">Install</a></li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#table-notation">Table notation</a></li>
    <li><a href="#faq">FAQ</a></li>
    <li><a href="#credits">Credits</a></li>
    <li><a href="#license">License</a></li>
  </ol>
</details>

## Features

* Price and 24H statistics updates for all BUSD/USDT pairs at Binance
  - Real-time price
  - Statistics update every 60 seconds
  - Choose to show BUSD or USDT or both pairs
  - UP/DOWN/BEAR/BULL pairs are excluded from the table
* Search the table with regex compatible patterns
* Full/mini table mode
  - When in mini mode only the top and bottom 15 pairs are shown
* Sort the table by a specific column 
* Cross platform
  - Windows, macOS and Linux ready.

<p align="right">(<a href="#top">back to top</a>)</p>

## Requirements

* `python 3.7.7+`

## Install

```sh
pip install termcs
```

### Run

```sh
termcs
```

<p align="right">(<a href="#top">back to top</a>)</p>

## Usage

To sort the table, simply click on the column header 

### Keybindings

* `f` - Full/mini table
* `/` - Search (by base asset name)
* `Esc` - Exit search mode
* `q` - Quit back to the terminal 

#### Pair control

* `b` - Show BUSD pairs only
* `t` - Show USDT pairs only
* `o` - Show both pairs
* `p` - Show/hide quote asset name

<p align="right">(<a href="#top">back to top</a>)</p>

## Table notation

| Column      | Description                                                                         |
|-------------|-------------------------------------------------------------------------------------|
| Price       | Current price (USD)                                                                 |
| Change      | The difference between the current price and the price 24 hours ago (percentage)    |
| High        | Highest price for the last 24 hours (USD)                                           |
| Low         | Lowest price for the last 24 hours (USD)                                            |
| High Change | The difference between the current price and the highest 24 hour price (percentage) |
| Low Change  | The difference between the current price and the lowest 24 hour price (percentage)  |
| Volume      | Trading pair volume for the last 24 hours                                           |

<p align="right">(<a href="#top">back to top</a>)</p>

## FAQ

#### Q: What's determine the asset pair when both pairs are presented in the table ?

A: The pair with the higher volume.

#### Q: Why there is a note "CHANGE PAIR RESTRICTION ENABLED" above the table ?

A: There is a limit of (approx) 27 times you can press the `b` / `t` / `o` keys in one minute, it is done in order to respect the API limit usage and avoid bans

[Read more](https://binance-docs.github.io/apidocs/spot/en/#limits)

<p align="right">(<a href="#top">back to top</a>)</p>

## Credits

This software uses the following open source packages:

- [textual](https://github.com/Textualize/textual)
- [requests](https://github.com/psf/requests)
- [websockets](https://github.com/python-websockets/websockets)

## License

MIT

<p align="right">(<a href="#top">back to top</a>)</p>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.4] - 2023-09-01

### Changed

- Use a websocket stream to get a real-time price

## [0.1.3] - 2023-08-25

### Fixed

- Updated weight regulation to the current (2023-08-25) Binance weight policies

### Changed

- Dropped binance-connector dependency

## [0.1.2] - 2023-04-18

### Fixed

- Fix footer keys highlighting https://github.com/Riyum/termcs/issues/4

## [0.1.1] - 2023-04-16

### Fixed

- Fix horizontal table misfit
- Improved table performance

## [0.1.0] - 2023-03-29

### Changed

- Overall increased performance due to the usage of a proper table widget from Textual

### Added

- Added the ability to sort the table by some column

### Removed

- Remove `stop` keybinding

## [0.0.2] - 2023-01-13

### Changed

- Use the right crypto terminology for the project

## [0.0.1] - 2023-01-11

- Initial release

[0.1.4]: https://github.com/Riyum/termcs/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/Riyum/termcs/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/Riyum/termcs/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/Riyum/termcs/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/Riyum/termcs/compare/v0.0.2...v0.1.0
[0.0.2]: https://github.com/Riyum/termcs/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/Riyum/termcs/releases/tag/v0.0.1



termcs/changeTable.py
from typing import List, Dict
from time import time
import re

from rich.text import Text
from textual.app import ComposeResult
from textual.geometry import Size
from textual.reactive import reactive
from textual.widgets import Static, DataTable, Label
from textual.containers import Vertical, Horizontal

# from textual.containers import Horizontal, Vertical, Container, Grid

from .worker import STATS_WEIGHT, TERMCS_WEIGHT, WEIGHT_LIMIT, Pair, RequestType, Worker
from .utils import RepeatedTimer


class ChangeTable(Static):
    """
    fetch market data from Binance into DataTable.
    child widgets updated by their respect threads that initialized in on_mount()
    """

    search_pattern = reactive("", init=False)
    full_table = reactive(False)
    show_pair = reactive(False)

    def __init__(self) -> None:
        super().__init__()
        self.worker = Worker(thresh=60)
        self.table = DataTable(zebra_stripes=True)
        self.initTable()

    def on_mount(self) -> None:
        self.tableStats_thread = RepeatedTimer(1, self.updateTableStats)
        # self.worker_thread = RepeatedTimer(3, self.updateTable)

    def compose(self) -> ComposeResult:
        yield Vertical(Label("", id="asset_count_label"), classes="label")
        yield Vertical(Label("", id="eta_label"), classes="label")
        yield Vertical(Label("", id="warning_label"), classes="label")
        yield Horizontal(self.table, id="table")

    def watch_search_pattern(self) -> None:
        self.table.clear(True)
        self.initTable(False)

    def initTable(self, req=True) -> None:
        """
        initialize an empty table

        Args:
            req(bool): if True fetch new data from the API, otherwise use previous cached data
        """
        self.col_keys = self.table.add_columns(
            *[
                "Pair" if self.show_pair else "Asset",
                "Price",
                "Change",
                "High",
                "Low",
                "High change",
                "Low change",
                "Volume",
            ]
        )

        if req:
            self.worker.getAll()

        table = list(self.worker.buff.values())
        table = table if self.full_table else [*table[:15], *table[-15:]]
        table = self.filt(table, self.search_pattern)
        table = self.sortAssetsList(table)

        for asset in table:
            row_asset = self.assetToRow(asset)
            self.table.add_row(*row_asset, key=asset["symbol"])

    # def fillTableWithRawData(self, table: List[Dict]) -> None:
    #     """fill an empty table with numbers"""
    #     self.table.clear(True)
    #     for asset in table:
    #         row = [
    #             asset["asset_name"],
    #             asset["price"],
    #             asset["change24"],
    #             asset["high"],
    #             asset["low"],
    #             asset["high_change"],
    #             asset["low_change"],
    #             asset["volume"],
    #         ]
    #         self.table.add_row(*row, key=asset["symbol"])

    def sortAssetsList(
        self, assets: List[Dict], key: str = "change24", rev: bool = True
    ) -> List[Dict]:
        return sorted(assets, key=lambda x: x[key], reverse=rev)

    def updateTableStats(self) -> None:
        """update the labels above the table"""

        """
        Since this function is invoked every second,
        it's worth checking whether the worker has faced any connectivity problems
        """
        if self.worker.kill:
            self.stop()
            self.app.exit("Connection problem ,check your internet, aborting...")

        if WEIGHT_LIMIT - TERMCS_WEIGHT - self.worker.used_weight < STATS_WEIGHT:
            self.query_one("#eta_label", Label).styles.margin = (0, 0)
            self.query_one("#warning_label", Label).styles.margin = (0, 0, 1, 0)
            self.query_one("#warning_label", Label).update(
                "!!! CHANGE PAIR RESTRICTION ENABLED !!!"
            )
        else:
            self.query_one("#warning_label", Label).styles.margin = (0, 0)
            self.query_one("#eta_label", Label).styles.margin = (0, 0, 1, 0)
            self.query_one("#warning_label", Label).update()

        self.query_one("#asset_count_label", Label).update(
            f"Assets: {self.worker.asset_count}"
        )

        eta = int(self.worker.update_time - time())
        self.query_one("#eta_label", Label).update(
            f"Update in: {eta if eta > 0 else 0}"
        )

    def updateTable(self) -> None:
        """update (an initialized) table with new data"""
        self.worker.getAll()

        if int(self.worker.update_time - time()) <= 0:
            """sort on stats update"""
            self.table.clear(True)
            self.initTable(False)
            return

        table = list(self.worker.buff.values())
        table = table if self.full_table else [*table[:15], *table[-15:]]
        table = self.filt(table, self.search_pattern)

        for asset in table:
            row_asset = self.assetToRow(asset)
            for i, val in enumerate(row_asset[1:]):
                self.table.update_cell(asset["symbol"], self.col_keys[i + 1], val)

    def assetToRow(self, asset: Dict) -> List:
        """
        Convert the asset data into a string or Rich text.
        return a list that will be used as a row for the DataTable.
        """
        name = asset["asset_name"]
        if self.show_pair:
            name = Text(asset["asset_name"])
            name.append(asset["symbol"][-4:], style="#fafab4 italic")

        change = Text(str(asset["change24"]))
        change.stylize(
            "green"
            if asset["change24"] > 0
            else ("red" if asset["change24"] < 0 else "")
        )

        if asset["price"] < 10**-4:
            price = f'{asset["price"]:.8f}'
            high = f'{asset["high"]:.8f}'
            low = f'{asset["low"]:.8f}'
        else:
            price = repr(asset["price"])
            high = repr(asset["high"])
            low = repr(asset["low"])

        low_change = repr(asset["low_change"])
        high_change = repr(asset["high_change"])

        r = [
            name,
            price,
            change,
            high,
            low,
            high_change,
            low_change,
            f'{asset["volume"]:,}',
        ]

        return r

    def filt(self, table: List[Dict], sp: str) -> List[Dict]:
        """Filter the table according to a specific pattern"""
        if not sp:
            return table
        try:
            pattern = re.compile(sp.upper())
        except re.error:
            return table

        return [asset for asset in table if pattern.search(asset["asset_name"])]

    def fullTable(self) -> None:
        self.full_table = not self.full_table
        self.table.clear(True)
        self.initTable(False)

    def setPairTo(self, p: Pair) -> None:
        self.worker.setPairTo(p)
        self.table.clear(True)
        self.initTable()

    def showPair(self) -> None:
        self.show_pair = not self.show_pair
        self.table.clear(True)
        self.initTable(False)

    def stop(self) -> None:
        self.active = False
        # self.worker_thread.stop()
        self.tableStats_thread.stop()

    def hasWeight(self) -> bool:
        """check if the stats request can be made"""
        return self.worker.hasWeightFor(RequestType.STATS, user=True)
